# Welcome to PrimerTK documentation

PrimerTK is a toolkit is a python package that has the capability to design standard primer sets, multiplex primer pools, and primer sets
around complex structural variants such as large deletions, insertions, translocation events, and inversions. 

PrimerTK is used in conjunction with [primer3](https://github.com/primer3-org/primer3) and [in silico PCR](https://hgwdev.gi.ucsc.edu/~kent/src/) to extract reference sequences by coordinate position, design primers, compare them with each other to check for pooling compatibility (if multiplexing), check for off-target products, and filter for the top results. It also has the ability to detect SNP's present in primers and annotate them in results.

# Table of Contents
1. [Introduction](#Introduction)
2. [Install](#Installation)
3. [Quickstart](#Quickstart)
4. [Standard Tutorial](#Standard-Primer-Generation)
5. [Multiplex Tutorial](#Multiplex-Primer-Generation)
6. [Deletions](#Deletions-Primer-Generation)
7. [Insertions](#Insertions-Primer-Generation )
0. [Test](#test.md)

## Introduction
This is an introduction to PrimerTK.

## Installation
Edit installer.sh then run to install primer3 and isPcr.
Also, `pip3 install primer_tk --user`

## Quickstart
This is a quick run through of the standard primer generation.

## Standard Primer Generation
This is the standard primer generation guide.

## Multiplex Primer Generation
This is the multiplex primer generation guide.

## Deletions Primer Generation
This is the guide to generate primers around large deletions.

## Insertions Primer Generation
This is the guide to generate primers around large insertions.
